runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/grpcclient/GrpcGetWorkerMetadataStreamTest.java

```diff
@@ -0,0 +1,348 @@
+
+    List<WorkerMetadataResponse.Endpoint> newDirectPathEndpoints =
+        Lists.newArrayList(
+            WorkerMetadataResponse.Endpoint.newBuilder()
+                .setEndpoint("newWindmillEndpoint")
+                .build());
+    Map<String, WorkerMetadataResponse.Endpoint> newGlobalDataEndpoints = new HashMap<>();
+    newGlobalDataEndpoints.put(
+        "new_global_data",
+        WorkerMetadataResponse.Endpoint.newBuilder().setEndpoint("newGlobalData").build());
+
+    WorkerMetadataResponse newWorkMetadataResponse =
+        WorkerMetadataResponse.newBuilder()
+            .setMetadataVersion(initialResponse.getMetadataVersion() + 1)
+            .addAllDirectPathEndpoints(newDirectPathEndpoints)
+            .putAllGlobalDataEndpoints(newGlobalDataEndpoints)
+            .build();
+
+    testStub.injectWorkerMetadata(newWorkMetadataResponse);
+
+    assertEquals(
+        newGlobalDataEndpoints.size(), testWindmillEndpointsConsumer.globalDataEndpoints.size());
+    testWindmillEndpointsConsumer.globalDataEndpoints.forEach(
+        (key, value) -> assertTrue(newGlobalDataEndpoints.containsKey(key)));
+    assertEquals(
+        newDirectPathEndpoints.size(), testWindmillEndpointsConsumer.windmillEndpoints.size());
+    assertTrue(
+        testWindmillEndpointsConsumer.windmillEndpoints.containsAll(
+            newDirectPathEndpoints.stream()
+                .map(WorkerMetadataResponse.Endpoint::getEndpoint)
+                .collect(Collectors.toList())));
+    assertEquals(newWorkMetadataResponse.getMetadataVersion(), metadataVersion.get());
+  }
+
+  @Test
+  public void testGetWorkerMetadata_doesNotConsumeResponseIfMetadataStale() {
+    WorkerMetadataResponse freshEndpoints =
+        WorkerMetadataResponse.newBuilder()
+            .setMetadataVersion(2)
+            .addAllDirectPathEndpoints(DIRECT_PATH_ENDPOINTS)
+            .putAllGlobalDataEndpoints(GLOBAL_DATA_ENDPOINTS)
+            .build();
+    TestWindmillEndpointsConsumer testWindmillEndpointsConsumer =
+        Mockito.spy(new TestWindmillEndpointsConsumer());
+    TestGetWorkMetadataRequestObserver requestObserver =
+        new TestGetWorkMetadataRequestObserver(freshEndpoints);
+    GetWorkerMetadataTestStub testStub = new GetWorkerMetadataTestStub(requestObserver);
+    AtomicLong metadataVersion = new AtomicLong(0);
+    stream = getWorkerMetadataStream(testStub, metadataVersion, testWindmillEndpointsConsumer);
+    List<WorkerMetadataResponse.Endpoint> staleDirectPathEndpoints =
+        Lists.newArrayList(
+            WorkerMetadataResponse.Endpoint.newBuilder()
+                .setEndpoint("staleWindmillEndpoint")
+                .build());
+    Map<String, WorkerMetadataResponse.Endpoint> staleGlobalDataEndpoints = new HashMap<>();
+    staleGlobalDataEndpoints.put(
+        "stale_global_data",
+        WorkerMetadataResponse.Endpoint.newBuilder().setEndpoint("staleGlobalData").build());
+
+    testStub.injectWorkerMetadata(
```

Comment:
see comment in code. It seems like this test should fail with the current logic.  It would also pass if this injection is somehow not reaching the stream so maybe some wiring is wrong?

Reply:
it failed (i didnt rerun it after i removed the locking)
passes now that the logic is fixed